[Turbopack Nextjs 16]: Pino - Cannot find module './transport-stream'

**Summary:** Discussions in this issue have diverged a bit, and there are mainly two cases:
1. **Turbopack + non-Vercel deployment.** Broken since v16.0.3 due to Turbopack starts to trace worker_threads (fixed #84766).
    - Errors (include but not limited to):
        - Parsing ecmascript source code failed (License file) - [comment1](https://github.com/vercel/next.js/issues/86099#issuecomment-3531120854)
        - Can't resolve 'why-is-node-running'/'tap'/etc. (require statement) - [comment2](https://github.com/vercel/next.js/issues/86099#issuecomment-3554743309), [comment3](https://github.com/vercel/next.js/issues/86099#issuecomment-3597186965)
    - Workaround:
        - Switch back to Webpack. `next build --webpack`
        - Or add `pino`, `thread-stream` and related pino transport target packages to `serverExternalPackages`, you might also need to add `thread-stream` as a dependency explicitly.
            - [comment4](https://github.com/vercel/next.js/issues/86099#issuecomment-3542407508) - Set `serverExternalPackages`
            - [comment5](https://github.com/vercel/next.js/issues/86099#issuecomment-3609795859), [comment6](https://github.com/vercel/next.js/issues/86099#issuecomment-3612899643) - Install `thread-stream`
            - [comment7](https://github.com/vercel/next.js/issues/86099#issuecomment-3613016995) - Use shim
            - [comment8](https://github.com/vercel/next.js/issues/86099#issuecomment-3613016995) - pin to specific versions due to conflicts
    - Fixed: since v16.1.0-canary.16 by #86375 and #86884
2. **Turbopack + Vercel deployement.** This is the original case of this issue. Turbopack never works in this setup, #84766 was fixed in v16.0.3, and now #86099 is waiting for a fix. The fixes above doesn't work for this case (see https://github.com/vercel/next.js/issues/86099#issuecomment-3629882719). The only workaround is to use Webpack `next build --webpack` with `serverExternalPackages`, see Notes below.

**Notes:**
From [comment9](https://github.com/vercel/next.js/issues/86099#issuecomment-3547917220)
> Pino is way too dynamic for Turbopack to be able to figure out all required files automatically: [pinojs/pino@7dd79a3/lib/transport.js#L94](https://github.com/pinojs/pino/blob/7dd79a3370f34e0c092dcbc2c6fb1748afcc1cdd/lib/transport.js#L94)

Pino itself has [a documentation](https://github.com/pinojs/pino/blob/main/docs/bundling.md) that talks about bundling, and there were already problems since Webpack + Vercel deployment, see [comment10](https://github.com/vercel/next.js/issues/86099#issuecomment-3549289622).

Also for in−depth details: https://github.com/pinojs/thread-stream/issues/184

---
### Link to the code that reproduces this issue

https://github.com/CHC383/nextjs-turbopack-pino-reproduction

### To Reproduce

1. Deploy the code to Vercel
2. Visit <url>/api/log
3. Check the log of the Vercel deployment

### Current vs. Expected behavior

## Current behavior

With Turbopack in NextJs v16, Pino worker thread throws the following error:

```
Uncaught Exception: Error: Cannot find module './transport-stream'
Require stack:
- /var/task/node_modules/.pnpm/pino@10.1.0/node_modules/pino/lib/worker.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1420:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1058:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1063:22)
    at Module._load (node:internal/modules/cjs/loader:1226:37)
    at TracingChannel.traceSync (node:diagnostics_channel:328:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:244:24)
    at Module.require (node:internal/modules/cjs/loader:1503:12)
    at a (/opt/rust/bytecode.js:2:1094)
    at Object.<anonymous> (/var/task/node_modules/.pnpm/pino@10.1.0/node_modules/pino/lib/worker.js:7:36)
    at Module.<anonymous> (/opt/rust/bytecode.js:2:1435)
```

## Expected behavior

Pino worker thread works without any exception, which is the behavior of using
WebPack.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #89-Ubuntu SMP PREEMPT_DYNAMIC Sat Oct 11 01:02:46 UTC 2025
  Available memory (MB): 15839
  Available CPU cores: 20
Binaries:
  Node: 24.11.1
  npm: 11.6.2
  Yarn: N/A
  pnpm: 10.24.0
Relevant Packages:
  next: 16.1.0-canary.16 // Latest available version is detected (16.1.0-canary.16).
  eslint-config-next: N/A
  react: 19.2.1
  react-dom: 19.2.1
  typescript: 5.9.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

### Additional context

@mischnic This is a follow up of #84766. It seems that with multiple targets, the `pino/lib/worker.js` is used and not traced by Turbopack properly. I have updated the reproduction code to trigger the new error.

I tested with the `serverExternalPackages` stopgap but it breaks.
On my end it's ignoring the exclusion of `thread-stream`

```
./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/test/helper.js:33:15
Module not found: Can't resolve 'why-is-node-running'
  31 |
  32 | if (process.env.SKIP_PROCESS_EXIT_CHECK !== 'true') {
> 33 |   const why = require('why-is-node-running')
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  34 |   setInterval(why, 10000).unref()
  35 | }
  36 |
```

Currently reverted to using webpack with this config:
```
webpack: (config) => {
    config.externals.push("pino-pretty", "lokijs", "encoding");
    return config;
  },
```

Having the same issue in 16.0.6

I have set `serverExternalPackages` as described above. I am using Turbopack. We host our app on Vercel, but this is breaking locally with `pnpm build`

```[1] ./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/test/string-limit.test.js:3:11
[1] Module not found: Can't resolve 'tap'
[1]   1 | 'use strict'
[1]   2 |
[1] > 3 | const t = require('tap')
[1]     |           ^^^^^^^^^^^^^^
[1]   4 |
[1]   5 | if (process.env.CI) {
[1]   6 |   t.skip('skip on CI')
[1]
[1]
[1]
[1] Import trace:
[1]   Client Component SSR:
[1]     ./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/test/string-limit.test.js [Client Component SSR]
[1]     ./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/index.js [Client Component SSR]
[1]     ./node_modules/.pnpm/pino@10.0.0/node_modules/pino/lib/transport.js [Client Component SSR]
[1]     ./node_modules/.pnpm/pino@10.0.0/node_modules/pino/pino.js [Client Component SSR]```

The fixes above that add pino related packages to `serverExternalPackages` doesn't fix it for me.

```shell
./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/test/transpiled.test.js:3:18
Module not found: Can't resolve 'tap'
  1 | 'use strict'
  2 |
> 3 | const { test } = require('tap')
    |                  ^^^^^^^^^^^^^^
  4 | const { join } = require('path')
  5 | const { file } = require('./helper')
  6 | const ThreadStream = require('..')



Import traces:
  App Route:
    ./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/test/transpiled.test.js
    ./node_modules/.pnpm/thread-stream@3.1.0/node_modules/thread-stream/index.js
    ./node_modules/.pnpm/pino@9.14.0/node_modules/pino/lib/transport.js
    ./node_modules/.pnpm/pino@9.14.0/node_modules/pino/pino.js
    ./node_modules/.pnpm/payload@3.64.0_graphql@16.12.0_typescript@5.7.2/node_modules/payload/dist/utilities/logger.js
```

facing same issue after upgrading to `v16.0.7`, though everything is fine on `v16.0.1`

```
./node_modules/.bun/thread-stream@3.1.0/node_modules/thread-stream/test/transpiled.test.js:3:18
Module not found: Can't resolve 'tap'
  1 | 'use strict'
  2 |
> 3 | const { test } = require('tap')
    |                  ^^^^^^^^^^^^^^
  4 | const { join } = require('path')
  5 | const { file } = require('./helper')
  6 | const ThreadStream = require('..')



Import trace:
  Client Component SSR:
    ./node_modules/.bun/thread-stream@3.1.0/node_modules/thread-stream/test/transpiled.test.js [Client Component SSR]
    ./node_modules/.bun/thread-stream@3.1.0/node_modules/thread-stream/index.js [Client Component SSR]
    ./node_modules/.bun/pino@10.0.0/node_modules/pino/lib/transport.js [Client Component SSR]
    ./node_modules/.bun/pino@10.0.0/node_modules/pino/pino.js [Client Component SSR]
    ./node_modules/.bun/@walletconnect+logger@3.0.0/node_modules/@walletconnect/logger/dist/index.es.js [Client Component SSR]
    ./node_modules/.bun/@walletconnect+utils@2.22.4+e8a3dd39f5b02c32/node_modules/@walletconnect/utils/dist/index.js [Client Component SSR]
    ./node_modules/.bun/@walletconnect+ethereum-provider@2.22.4+7b56cfec410038ca/node_modules/@walletconnect/ethereum-provider/dist/index.js [Client Component SSR]
    ./node_modules/.bun/@privy-io+react-auth@3.7.0+d827b7e707306ea4/node_modules/@privy-io/react-auth/dist/esm/EmbeddedWalletConnectingScreen-D50stmJi.mjs [Client Component SSR]
    ./node_modules/.bun/@privy-io+react-auth@3.7.0+d827b7e707306ea4/node_modules/@privy-io/react-auth/dist/esm/index.mjs [Client Component SSR]
```

--- a/docs/01-app/03-api-reference/05-config/01-next-config-js/serverExternalPackages.mdx
+++ b/docs/01-app/03-api-reference/05-config/01-next-config-js/serverExternalPackages.mdx
@@ -73,6 +73,9 @@ Next.js includes a [short list of popular packages](https://github.com/vercel/ne
 - `onnxruntime-node`
 - `oslo`
 - `pg`
+- `pino`
+- `pino-pretty`
+- `pino-roll`
 - `playwright`
 - `playwright-core`
 - `postcss`
@@ -86,8 +89,9 @@ Next.js includes a [short list of popular packages](https://github.com/vercel/ne
 - `sharp`
 - `shiki`
 - `sqlite3`
-- `ts-node`
+- `thread-stream`
 - `ts-morph`
+- `ts-node`
 - `typescript`
 - `vscode-oniguruma`
 - `webpack`

--- a/docs/02-pages/04-api-reference/04-config/01-next-config-js/serverExternalPackages.mdx
+++ b/docs/02-pages/04-api-reference/04-config/01-next-config-js/serverExternalPackages.mdx
@@ -73,6 +73,9 @@ Next.js includes a [short list of popular packages](https://github.com/vercel/ne
 - `onnxruntime-node`
 - `oslo`
 - `pg`
+- `pino`
+- `pino-pretty`
+- `pino-roll`
 - `playwright`
 - `playwright-core`
 - `postcss`
@@ -86,8 +89,9 @@ Next.js includes a [short list of popular packages](https://github.com/vercel/ne
 - `sharp`
 - `shiki`
 - `sqlite3`
-- `ts-node`
+- `thread-stream`
 - `ts-morph`
+- `ts-node`
 - `typescript`
 - `vscode-oniguruma`
 - `webpack`

--- a/packages/next/src/lib/server-external-packages.jsonc
+++ b/packages/next/src/lib/server-external-packages.jsonc
@@ -70,6 +70,14 @@
   "postcss",
   "prettier",
   "prisma",
+  // pino@10.1.0/lib/transport.js passes `join(__dirname, 'worker.js')` to thread-stream which ends
+  // up `require`ing that filepath, leading to
+  // `Cannot find module '/ROOT/node_modules/.pnpm/pino@10.1.0/node_modules/pino/lib/worker.js'`
+  "pino",
+  // A "plugin" that is passed to pino as a string (and is then required in a worker)
+  "pino-pretty",
+  // A "plugin" that is passed to pino as a string (and is then required in a worker)
+  "pino-roll",
   "puppeteer",
   "puppeteer-core",
   "ravendb",
@@ -80,6 +88,9 @@
   "sharp",
   "shiki",
   "sqlite3",
+  // dynamic `new Worker` call in thread-stream@3.1.0/index.js which pulls in various non-code files
+  // as well (trying to parse README and invalid test files)
+  "thread-stream",
   "ts-node",
   // Slow to compile because it pulls in typescript
   "ts-morph",

--- a/packages/next/src/lib/server-external-packages.jsonc
+++ b/packages/next/src/lib/server-external-packages.jsonc
@@ -65,11 +65,6 @@
   "onnxruntime-node",
   "oslo",
   "pg",
-  "playwright",
-  "playwright-core",
-  "postcss",
-  "prettier",
-  "prisma",
   // pino@10.1.0/lib/transport.js passes `join(__dirname, 'worker.js')` to thread-stream which ends
   // up `require`ing that filepath, leading to
   // `Cannot find module '/ROOT/node_modules/.pnpm/pino@10.1.0/node_modules/pino/lib/worker.js'`
@@ -78,6 +73,11 @@
   "pino-pretty",
   // A "plugin" that is passed to pino as a string (and is then required in a worker)
   "pino-roll",
+  "playwright",
+  "playwright-core",
+  "postcss",
+  "prettier",
+  "prisma",
   "puppeteer",
   "puppeteer-core",
   "ravendb",
@@ -91,9 +91,9 @@
   // dynamic `new Worker` call in thread-stream@3.1.0/index.js which pulls in various non-code files
   // as well (trying to parse README and invalid test files)
   "thread-stream",
-  "ts-node",
   // Slow to compile because it pulls in typescript
   "ts-morph",
+  "ts-node",
   "typescript",
   "vscode-oniguruma",
   "webpack",

